cirq/contrib/quimb/mps_simulator.py

```diff
@@ -231,92 +235,243 @@ def sample(
         return np.array(measurements, dtype=int)
 
 
+def _sum_reduce(M, ind):
+    # TODO(tonybruguier): Use M.sum_reduce(ind, inplace=True) once
+    # version 2.0 of Quimb is released.
+    summer = qtn.Tensor([1.0] * M.ind_size(ind), inds=(ind,))
+    return _make_tensor_if_scalar(M @ summer)
+
+
+def _make_tensor_if_scalar(x):
+    # Sometimes, we get a scalar, but we would rather keep it as tensor.
+    if not isinstance(x, qtn.Tensor):
+        return qtn.Tensor(x)
+    return x
+
+
 @value.value_equality
 class MPSState:
     """A state of the MPS simulation."""
 
-    def __init__(self, qubit_map, initial_state=0):
+    def __init__(self, qubit_map, rsum2_cutoff, initial_state=0):
         self.qubit_map = qubit_map
         self.M = []
-        for qubit in qubit_map.keys():
+
+        # The order of the qubits matters, because the state |01> is different from |10>. Since
+        # Qimb uses strings to name tensor indices, we want to be able to sort them too. If we are
+        # working with, say, 123 qubits then we want qubit 3 to come before qubit 100, but then
+        # we want write the string '003' which comes before '100' in lexicographic order. The code
+        # below is just simple string formatting.
+        max_num_digits = len('%d' % (max(qubit_map.values())))
+        self.format_i = 'i_%%.%dd' % (max_num_digits)
+        self.format_mu = 'mu_%%.%dd_%%.%dd' % (max_num_digits, max_num_digits)
+
+        for qubit in reversed(qubit_map.keys()):
             d = qubit.dimension
-            x = np.zeros(
-                (
-                    1,
-                    1,
-                    d,
-                )
-            )
-            x[0, 0, (initial_state % d)] = 1.0
-            self.M.append(x)
+            x = np.zeros(d)
+            x[initial_state % d] = 1.0
+
+            i = qubit_map[qubit]
+            self.M.append(qtn.Tensor(x, inds=(self.i_str(i),)))
             initial_state = initial_state // d
         self.M = self.M[::-1]
-        self.threshold = 1e-3
+        self.rsum2_cutoff = rsum2_cutoff
+        self.num_svd_splits = 0
+
+    def i_str(self, i):
+        return self.format_i % (i)
+
+    def mu_str(self, i, j):
+        # By convention, the lower index is always the first.
+        i, j = min(i, j), max(i, j)
```

Comment:
could split into two lines for more readability

Reply:
Done.